voxxed-microservices-2018/module-01.adoc

```diff
@@ -90,128 +127,86 @@ The command output should be similar to:
 
 [source,sh]
 NAME                                          READY     STATUS    RESTARTS   AGE
-strimzi-cluster-controller-2044197322-lzrvr   1/1       Running   0          3m
+strimzi-cluster-operator-2044197322-lzrvr   1/1       Running   0          3m
 
 Next, install the Strimzi templates.
-The Cluster Controller related templates contain predefined config maps for easily deploying clusters (for Kafka Connect as well).
+The Cluster Operator related templates contain predefined resources for easily deploying clusters (for Kafka Connect as well).
 
 [source, sh]
-$ oc create -f examples/templates/cluster-controller
-template "strimzi-connect-s2i" created
-template "strimzi-connect" created
-template "strimzi-ephemeral" created
-template "strimzi-persistent" created
-
-The Topic Controller related templates contain predefined config map for easily creating a new Kafka topic.
+$ oc apply -f examples/templates/cluster-operator -n voxxed
+template.template.openshift.io "strimzi-connect-s2i" created
+template.template.openshift.io "strimzi-connect" created
+template.template.openshift.io "strimzi-ephemeral" created
+template.template.openshift.io "strimzi-persistent" created
 
-[source,sh]
-$ oc create -f examples/templates/topic-controller
-template "strimzi-topic" created
+Now you can deploy a Kafka cluster.
+For this lab, we will use a template, but you could create a resource file from scratch as well.
 
-Now you can deploy a Kafka cluster by creating a config map.
-For this lab, we will use a template, but you can create and deploy a regular config map as well.
+[source, sh]
+$ oc apply -f examples/kafka/kafka-ephemeral.yaml
```

Comment:
Replaxce with `oc process strimzi-ephemeral -p ZOOKEEPER_NODE_COUNT=1 | oc apply -f -` to deploy only one ZK

Reply:
Ah, that's cool, 👍 I was about to remove the step about adding the templates before because it seemed not needed with the `apply`, but now it makes sense again. Thanks, @jpechane!

Reply:
> The templates in 0.7.0 IMHO do not have the metrics enabled - so that will for sure not work with Prometheus.

Ok, so what will we have to do then to enable metrics?

> With ephemeral storage you need one restart of the pod because of memory or something and thinks can quickly go to hell.

Hum, that sounds a bit vague :) I tried what @jpechane suggested and it works. Not sure what to do now... The thing is we'd like to stay conservative on the resource usage side, so to burden people's machines not more than needed (Minishift or similar local set-up might not handle it well otherwise).

Reply:
> Ok, so what will we have to do then to enable metrics?

To enable the metrics you need to have the `Kafka.spec.kafka.metrics` and `Kafka.spec.zookeeper.metrics` fields in the custom resource like [here](https://github.com/strimzi/strimzi-kafka-operator/blob/1d7939c8b19f626644b1a050a4899ef16d54cc79/examples/kafka/kafka-ephemeral.yaml#L23). But this is missing in the [template](https://github.com/strimzi/strimzi-kafka-operator/blob/0.7.0/examples/templates/cluster-operator/ephemeral-template.yaml).

> Hum, that sounds a bit vague :) I tried what @jpechane suggested and it works. Not sure what to do now... The thing is we'd like to stay conservative on the resource usage side, so to burden people's machines not more than needed (Minishift or similar local set-up might not handle it well otherwise).

The ephemeral storage means it is using `emptyDir` volume. When the pod restarts, it typically looses all the data and starts with empty directory again. But Kafka doesn't like it that suddenly the Zookeeper starts from 0 and doesn't have any data which were there previously.

To avoid this when you use just one Zookeeper node, you should use the persistent storage. On Minishift or `oc cluster up` this is normally emulated and works out of the box. With persistent storage, when the Zookepeer pod restarts, it keeps all the files it created and thanks to that works fine even with single node.

So if you want the metrics, I would recommend you to use this [YAML file](https://github.com/strimzi/strimzi-kafka-operator/blob/0.7.0/examples/kafka/kafka-persistent.yaml) with `oc apply -f` and just ask the users to change the number of replicas.

Reply:
> To enable the metrics you need to have the Kafka.spec.kafka.metrics and Kafka.spec.zookeeper.metrics fields in the custom resource like here. But this is missing in the template.

I see. FWIW, I used that example resource you linked before and it didn't work either. I'll try and see whether the stuff even is in Prometheus or not. I was essentially following the 0.7.0 instructions (which don't use the template), but something still seems off.

> But Kafka doesn't like it that suddenly the Zookeeper starts from 0 and doesn't have any data which were there previously.

Ok, gotcha. But should it be fine then if we start with a single ZK from the beginning? If not, I can give the persistent setting a try; hoped I could avoid it so far :)

Reply:
> But should it be fine then if we start with a single ZK from the beginning? If not, I can give the persistent setting a try; hoped I could avoid it so far :)

It will be fine until whatever accidental restart happens to the Zookeeper pod. So in most case it should work.

Reply:
Ok, so would the change to the persistent approach (using `oc cluster up`) just mean we'd use the "persistent" custom example resource / template instead of the "ephermal" one? And how persistent is this, i.e. will state survive `oc cluster down` and `up`, or will the latter start with a clean slate?

Reply:
Yes, it means just using different example to start with. It doesn't survive `oc cluster down` and `oc cluster up` becasue that deletres all openshift resources (the data would actually survive, but you will be unable to use them easily), but it would survive `minikube stop` and `minikube start` for example.